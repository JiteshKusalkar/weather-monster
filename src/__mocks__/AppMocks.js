export const validProps = {
  cities: [
    {
      id: 1283240,
      name: 'Kathmandu',
      main: {
        temp: 57.2,
        temp_max: 57.2,
        temp_min: 57.2
      },
      weather: [
        {
          description: 'mist',
          icon: '50d',
          id: 701,
          main: 'Mist'
        }
      ]
    }
  ],
  deleteCity: jest.fn(),
  fetchCurrentWeatherById: jest.fn()
};

export const emptyProps = {
  cities: [],
  deleteCity: () => null,
  fetchCurrentWeatherById: () => null
};
