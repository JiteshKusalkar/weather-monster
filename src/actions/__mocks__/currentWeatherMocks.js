import { FETCH_CURRENT_WEATHER_REQUEST } from "../currentWeather";

export const response = [
  { type: FETCH_CURRENT_WEATHER_REQUEST },
  {
    payload: {
      base: 'stations',
      clouds: { all: 75 },
      cod: 200,
      coord: { lat: 27.72, lon: 85.32 },
      dt: 1576927921,
      id: 1283240,
      main: {
        feels_like: 12.66,
        humidity: 62,
        pressure: 1016,
        temp: 15,
        temp_max: 15,
        temp_min: 15
      },
      name: 'Kathmandu',
      sys: {
        country: 'NP',
        id: 9201,
        sunrise: 1576890297,
        sunset: 1576927681,
        type: 1
      },
      timezone: 20700,
      visibility: 7000,
      weather: [
        { description: 'broken clouds', icon: '04n', id: 803, main: 'Clouds' }
      ],
      wind: { deg: 270, speed: 2.6 }
    },
    type: 'ADD_CITY'
  }
];
