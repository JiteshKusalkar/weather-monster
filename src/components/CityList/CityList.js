import styled from 'styled-components';
import React from 'react';

const unit = process.env.REACT_APP_UNITS === 'imperial' ? 'C' : 'F';

const CityList = ({ cities, onDelete }) => {
  const renderWeatherCard = city => {
    const {
      name,
      main: { temp, temp_min: min, temp_max: max },
      weather
    } = city;

    return (
      <WeatherCard key={city.id} onClick={onDelete(city.id)}>
        <CityName>{name}</CityName>
        <TemperatureSummary>
          {weather.map(elem => (
            <WeatherIcon key={`icon-${name}-${elem.id}`} icon={elem.icon} />
          ))}
          <CurrentTemperature>{temp}</CurrentTemperature>
          <Unit size='lg'>{unit}</Unit>
        </TemperatureSummary>
        <Divider />
        <Temperature>
          <Label>Min.</Label>
          <Value>{min}</Value>
          <Unit size='sm'>{unit}</Unit>
        </Temperature>
        <Temperature>
          <Label>Max.</Label>
          <Value>{max}</Value>
          <Unit size='sm'>{unit}</Unit>
        </Temperature>
      </WeatherCard>
    );
  };

  return cities && <Wrapper>{cities.map(renderWeatherCard)}</Wrapper>;
};

export default CityList;

const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: 30px;
`;

const WeatherCard = styled.div`
  box-shadow: 0 6px 7px 0 rgba(0, 151, 182, 0.3);
  padding: 20px;
  background-color: #0097b6;
  color: #ffffff;
  width: 100%;
  margin: 0 20px 20px 0;
  border-radius: 5px;
  cursor: pointer;

  @media (min-width: 1024px) {
    width: calc((100% / 3) - 60px);
  }
`;

const TemperatureSummary = styled.div`
  display: flex;
  justify-content: center;
`;

const CurrentTemperature = styled.div`
  font-size: 60px;
  margin: 0;

  &:after {
    content: '°';
  }
`;

const Unit = styled.span`
  font-size: ${props => `${props.size === 'lg' ? '40px' : '20px'}`};
  margin: 0;
  vertical-align: top;
`;

const WeatherIcon = styled.div`
  width: 45px;
  height: 45px;
  background: ${props => `
    url(http://openweathermap.org/img/w/${props.icon}.png)
    repeat scroll 0% 0%
    transparent`};
`;

const CityName = styled.div`
  padding: 10px;
  text-align: center;
  font-size: 20px;
`;

const Divider = styled.div`
  width: 30px;
  border: 2px solid #ffffff;
  margin: 20px auto;
`;

const Temperature = styled.div`
  width: 50%;
  display: inline-block;
  text-align: center;
`;

const Value = styled.span`
  font-size: 30px;
  margin: 0;

  &:after {
    content: '°';
  }
`;

const Label = styled.div``;
