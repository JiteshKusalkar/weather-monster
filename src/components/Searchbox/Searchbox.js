import styled from 'styled-components';
import React, { useState } from 'react';
import Autosuggest from 'react-autosuggest';

import data from '../../data/weather.json';
import { escape } from '../../utils/common.js';

const Searchbox = ({ id, onSuggestionSelect }) => {
  const cities = data.map(cityData => ({
    name: cityData.city.name,
    id: cityData.city.id
  }));

  const [value, setValue] = useState('');
  const [suggestions, setSuggestions] = useState(cities);

  const getSuggestions = value => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;

    return inputLength === 0
      ? cities
      : cities.filter(city => city.name.toLowerCase().includes(inputValue));
  };

  // To empty the input string on suggestion click
  const getSuggestionValue = suggestion => '';

  const renderSuggestion = suggestion => {
    let suggestionLabel = suggestion.name;
    const queryRegex = new RegExp(escape(value), 'ig');
    suggestionLabel = suggestionLabel.replace(queryRegex, '<b>$&</b>');

    return <Suggestion dangerouslySetInnerHTML={{ __html: suggestionLabel }} />;
  };

  const onChange = (event, { newValue }) => {
    setValue(newValue);
  };

  const onSuggestionsFetchRequested = ({ value }) => {
    setSuggestions(getSuggestions(value));
  };

  const onSuggestionsClearRequested = () => {
    setSuggestions(cities);
  };

  const onSuggestionSelected = (event, { suggestion }) => {
    onSuggestionSelect(suggestion);
  };

  const inputProps = {
    placeholder: 'Type the name of a city',
    value,
    onChange
  };

  return (
    <Wrapper>
      <Autosuggest
        id={id}
        suggestions={suggestions}
        onSuggestionsFetchRequested={onSuggestionsFetchRequested}
        onSuggestionsClearRequested={onSuggestionsClearRequested}
        getSuggestionValue={getSuggestionValue}
        renderSuggestion={renderSuggestion}
        onSuggestionSelected={onSuggestionSelected}
        shouldRenderSuggestions={value => true}
        focusInputOnSuggestionClick={false}
        inputProps={inputProps}
      />
    </Wrapper>
  );
};

export default Searchbox;

const Wrapper = styled.div`
  padding: 10px;

  .react-autosuggest {
    &__container {
      width: 100%;
      margin: 0 auto;
      position: relative;
      padding-right: 20px;
      
      @media (min-width: 1024px) {
        width: 300px;
      }
    }

    &__suggestions-container {
      position: absolute;
      width: 100%;
      box-shadow: 0 6px 7px 0 rgba(0, 0, 0, 0.1);
      max-height: 200px;
      overflow: auto;
      background-color: #ffffff;
    }
  }

  ul {
    padding: 0;
    margin: 0;
    list-style: none;

    li {
      padding: 10px;
      border-top: 1px solid #e5e5e5;
      cursor: pointer;
    }
  }
`;

const Suggestion = styled.div``;
