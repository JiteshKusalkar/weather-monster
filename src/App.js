import styled from 'styled-components';
import React from 'react';
import { connect } from 'react-redux';
import { createGlobalStyle } from 'styled-components';

import { deleteCity, fetchCurrentWeatherById } from './actions/currentWeather';
import Searchbox from './components/Searchbox';
import CityList from './components/CityList';

const App = props => {
  const onSuggestionSelect = suggestion => {
    props.fetchCurrentWeatherById(suggestion.id);
  };

  const onDeleteCity = id => e => props.deleteCity(id);

  return (
    <Wrapper>
      <GlobalStyle />
      <Header>Weather Monster</Header>
      <Searchbox id='city-search' onSuggestionSelect={onSuggestionSelect} />
      <CityList cities={props.cities} onDelete={onDeleteCity} />
    </Wrapper>
  );
};

const mapStateToProps = state => ({
  cities: state.currentWeather.cities
});

const mapDispatchToProps = dispatch => ({
  fetchCurrentWeatherById: id => dispatch(fetchCurrentWeatherById(id)),
  deleteCity: id => dispatch(deleteCity(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);

const GlobalStyle = createGlobalStyle`
  body {
    margin: 0;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
      'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
      sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

  input {
    width: 100%;
    outline: none;
    border-radius: 5px;
    border: 1px solid #ccc;
    padding: 10px;
    line-height: 1;
  }
`;

const Header = styled.h1`
  font-size: 40px;
  font-weight: 400;
  text-align: center;
`;

const Wrapper = styled.div`
  margin-top: 10%;
`;
