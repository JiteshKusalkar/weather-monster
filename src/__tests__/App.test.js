import { shallow } from 'enzyme';
import React from 'react';
import unroll from 'unroll';

import CityList from '../components/CityList';
import { WeatherCard } from '../components/CityList/CityList';
import Searchbox from '../components/Searchbox';
import { emptyProps, validProps } from '../__mocks__/AppMocks';
import { App } from '../App';

const shallowInstance = props => shallow(<App {...props} />);
unroll.use(it);

describe('Snapshot: App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowInstance(emptyProps);
  });

  unroll(
    'should match the snapshot for #propType',
    (done, args) => {
      wrapper.setProps(args.value);
      expect(wrapper).toMatchSnapshot();
      done();
    },
    [
      ['propType', 'value'],
      ['valid props', validProps],
      ['empty props', emptyProps]
    ]
  );
});

// Unit tests
describe('Unit: App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowInstance(validProps);
  });

  it('should delete city by id', () => {
    wrapper
      .find(CityList)
      .dive()
      .find(WeatherCard)
      .simulate('click', 'evt');

    expect(validProps.deleteCity).toHaveBeenCalledWith(1283240);
  });

  it('should fetch current weather by id on suggestion select', () => {
    const suggestion = validProps.cities[0];
    wrapper.find(Searchbox).simulate('suggestionSelect', suggestion);

    expect(validProps.fetchCurrentWeatherById).toHaveBeenCalledWith(suggestion.id);
  });
});
