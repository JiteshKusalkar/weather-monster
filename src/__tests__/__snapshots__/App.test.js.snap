// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: App should match the snapshot for "empty props" 1`] = `ShallowWrapper {}`;

exports[`Snapshot: App should match the snapshot for "valid props" 1`] = `ShallowWrapper {}`;
